File: server/rails/mpesa.js
// ============================================================================
// M-Pesa Daraja API Payment Rail Connector
// Implements STK Push (Lipa Na M-Pesa), C2B/STK Callback, and B2C Payouts
// Settles directly into the double-entry accounting ledger
// ============================================================================
const crypto = require('crypto');
const db = require('../db');
const { recordIncome, recordSpending, getAccountBalance } = require('../ledger/ledger-core');

// Standard exchange rate for settlement: 1 USD ~ 130 KES (configurable)
const KES_TO_USD_RATE = 1 / 130.0;

function ensureMpesaAccount(userId) {
  let mpesaAccount = db.get(
    "SELECT id, name, currency FROM accounts WHERE user_id = ? AND (name LIKE '%M-Pesa%' OR id = 'mpesa_wallet')",
    [userId]
  );

  if (!mpesaAccount) {
    const accountId = 'acc_mpesa_' + crypto.randomBytes(4).toString('hex');
    db.run(
      `INSERT INTO accounts (id, user_id, name, type, account_class, currency, institution, initial_balance)
       VALUES (?, ?, 'M-Pesa Mobile Wallet', 'cash', 'asset', 'KES', 'Safaricom M-Pesa', 0)`,
      [accountId, userId]
    );
    mpesaAccount = { id: accountId, name: 'M-Pesa Mobile Wallet', currency: 'KES' };
  }

  return mpesaAccount;
}

/**
 * Initiates an M-Pesa STK Push (Lipa Na M-Pesa Online prompt) to user's phone
 */
function initiateStkPush({ userId, phoneNumber, amountKes, description = 'Lumen Wallet Top-up' }) {
  const numericAmount = Math.round(Number(amountKes));
  if (isNaN(numericAmount) || numericAmount <= 0) {
    throw new Error('M-Pesa amount must be a positive integer in KES');
  }

  // Format phone number: 2547XXXXXXXX or 07XXXXXXXX
  let cleanPhone = phoneNumber.replace(/[^0-9]/g, '');
  if (cleanPhone.startsWith('0')) {
    cleanPhone = '254' + cleanPhone.substring(1);
  } else if (!cleanPhone.startsWith('254')) {
    cleanPhone = '254' + cleanPhone;
  }

  if (cleanPhone.length !== 12) {
    throw new Error('Invalid Kenyan phone number format. Expected e.g. 0712345678 or 254712345678');
  }

  const mpesaAccount = ensureMpesaAccount(userId);
  const checkoutRequestId = 'ws_CO_' + Date.now() + '_' + crypto.randomBytes(4).toString('hex');
  const merchantRequestId = 'mr_' + crypto.randomBytes(6).toString('hex');
  const railTxId = 'rail_' + crypto.randomUUID();

  // Record pending inbound rail transaction
  db.run(
    `INSERT INTO payment_rails_transactions (id, user_id, rail, direction, status, amount, currency, reference, phone_or_account, metadata_json)
     VALUES (?, ?, 'mpesa', 'inflow', 'pending', ?, 'KES', ?, ?, ?)`,
    [
      railTxId,
      userId,
      numericAmount,
      checkoutRequestId,
      cleanPhone,
      JSON.stringify({
        description,
        merchantRequestId,
        accountId: mpesaAccount.id,
        convertedUsd: Number((numericAmount * KES_TO_USD_RATE).toFixed(2))
      })
    ]
  );

  return {
    success: true,
    checkoutRequestId,
    merchantRequestId,
    railTxId,
    phoneNumber: cleanPhone,
    amountKes: numericAmount,
    status: 'pending',
    message: `STK push prompt dispatched to ${cleanPhone}. Enter M-Pesa PIN on handset to complete deposit.`
  };
}

/**
 * Handles Daraja STK Push Callback Webhook
 * Upon successful PIN entry, credits user's M-Pesa ledger account atomically
 */
function handleMpesaCallback(payload) {
  const stkCallback = payload?.Body?.stkCallback || payload;
  if (!stkCallback) {
    throw new Error('Invalid M-Pesa callback body structure');
  }

  const checkoutRequestId = stkCallback.CheckoutRequestID;
  const resultCode = stkCallback.ResultCode;
  const resultDesc = stkCallback.ResultDesc;

  const railTx = db.get(
    "SELECT * FROM payment_rails_transactions WHERE reference = ? AND rail = 'mpesa'",
    [checkoutRequestId]
  );

  if (!railTx) {
    throw new Error(`No pending rail transaction found for CheckoutRequestID: ${checkoutRequestId}`);
  }

  if (railTx.status === 'completed') {
    return { success: true, message: 'Transaction already completed and reconciled' };
  }

  let metadata = {};
  try { metadata = JSON.parse(railTx.metadata_json); } catch (e) {}

  if (resultCode === 0) {
    // Payment Successful! Extract callback metadata items
    const callbackMetadata = stkCallback.CallbackMetadata?.Item || [];
    let receiptNumber = 'MPESA_' + crypto.randomBytes(4).toString('hex').toUpperCase();
    let transactionDate = new Date().toISOString();
    let paidAmount = railTx.amount;

    for (const item of callbackMetadata) {
      if (item.Name === 'MpesaReceiptNumber') receiptNumber = item.Value;
      if (item.Name === 'TransactionDate') transactionDate = String(item.Value);
      if (item.Name === 'Amount') paidAmount = Number(item.Value);
    }

    const mpesaAccount = ensureMpesaAccount(railTx.user_id);
    const usdAmount = Number((paidAmount * KES_TO_USD_RATE).toFixed(2));

    // Atomically settle in double-entry ledger:
    // Debit: asset:mpesa_account (increasing funds)
    // Credit: revenue:M-Pesa Deposit
    const ledgerResult = recordIncome({
      userId: railTx.user_id,
      accountId: mpesaAccount.id,
      amount: usdAmount,
      category: 'M-Pesa Deposit',
      payee: `M-Pesa Inward (${receiptNumber})`,
      referenceId: receiptNumber,
      icon: 'smartphone',
      tone: 'teal'
    });

    db.run(
      `UPDATE payment_rails_transactions
       SET status = 'completed', reference = ?, journal_entry_id = ?, updated_at = datetime('now')
       WHERE id = ?`,
      [receiptNumber, ledgerResult.journalEntryId, railTx.id]
    );

    return {
      success: true,
      status: 'completed',
      receiptNumber,
      amountKes: paidAmount,
      creditedUsd: usdAmount,
      message: `M-Pesa payment ${receiptNumber} confirmed and booked to ledger.`
    };
  } else {
    // Cancelled or Failed on handset
    db.run(
      `UPDATE payment_rails_transactions
       SET status = 'failed', metadata_json = ?, updated_at = datetime('now')
       WHERE id = ?`,
      [JSON.stringify({ ...metadata, resultDesc, resultCode }), railTx.id]
    );

    return {
      success: false,
      status: 'failed',
      resultCode,
      resultDesc
    };
  }
}

/**
 * Sends a B2C Payout to any M-Pesa phone number (Withdrawal / Outflow)
 */
function sendB2CPayout({ userId, recipientPhone, amountKes, description = 'Lumen Payout' }) {
  const numericAmount = Math.round(Number(amountKes));
  if (isNaN(numericAmount) || numericAmount <= 0) {
    throw new Error('Payout amount must be a positive integer in KES');
  }

  let cleanPhone = recipientPhone.replace(/[^0-9]/g, '');
  if (cleanPhone.startsWith('0')) cleanPhone = '254' + cleanPhone.substring(1);
  else if (!cleanPhone.startsWith('254')) cleanPhone = '254' + cleanPhone;

  const mpesaAccount = ensureMpesaAccount(userId);
  const usdAmount = Number((numericAmount * KES_TO_USD_RATE).toFixed(2));

  // Verify user has sufficient funds in M-Pesa account
  const currentBalance = getAccountBalance(mpesaAccount.id);
  if (currentBalance < usdAmount) {
    throw new Error(`Insufficient funds in M-Pesa account. Available: $${currentBalance.toFixed(2)} (approx KES ${(currentBalance / KES_TO_USD_RATE).toFixed(0)}), Requested: KES ${numericAmount} ($${usdAmount})`);
  }

  const receiptNumber = 'B2C_' + crypto.randomBytes(5).toString('hex').toUpperCase();
  const railTxId = 'rail_' + crypto.randomUUID();

  // Atomically debit M-Pesa account in double-entry ledger:
  // Debit: expense:M-Pesa Payout
  // Credit: asset:mpesa_account (decreasing funds)
  const ledgerResult = recordSpending({
    userId,
    accountId: mpesaAccount.id,
    amount: usdAmount,
    category: 'M-Pesa Payout',
    payee: `M-Pesa Outward to ${cleanPhone}`,
    referenceId: receiptNumber,
    icon: 'arrow-up-right',
    tone: 'orange'
  });

  db.run(
    `INSERT INTO payment_rails_transactions (id, user_id, rail, direction, status, amount, currency, reference, phone_or_account, metadata_json, journal_entry_id)
     VALUES (?, ?, 'mpesa', 'outflow', 'completed', ?, 'KES', ?, ?, ?, ?)`,
    [
      railTxId,
      userId,
      numericAmount,
      receiptNumber,
      cleanPhone,
      JSON.stringify({ description, convertedUsd: usdAmount }),
      ledgerResult.journalEntryId
    ]
  );

  return {
    success: true,
    status: 'completed',
    receiptNumber,
    recipientPhone: cleanPhone,
    amountKes: numericAmount,
    debitedUsd: usdAmount,
    newBalanceUsd: getAccountBalance(mpesaAccount.id),
    message: `KES ${numericAmount.toLocaleString()} successfully sent to ${cleanPhone} via M-Pesa B2C.`
  };
}

module.exports = {
  KES_TO_USD_RATE,
  ensureMpesaAccount,
  initiateStkPush,
  handleMpesaCallback,
  sendB2CPayout
};

File: src/rails/mpesa.rs
// ============================================================================
// M-Pesa Payment Rail Connector in Rust
// Daraja STK Push, Callback handling, and B2C Payouts with ledger integration
// ============================================================================
use crate::ledger::core::{get_account_balance, record_income, record_spending, TransactionResult};
use rusqlite::{params, Connection};
use rust_decimal::Decimal;
use rust_decimal_macros::dec;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

pub const KES_TO_USD_RATE: Decimal = dec!(0.00769230769); // 1 / 130.0

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StkPushResult {
    pub success: bool,
    pub checkout_request_id: String,
    pub merchant_request_id: String,
    pub rail_tx_id: String,
    pub phone_number: String,
    pub amount_kes: i64,
    pub status: String,
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PayoutResult {
    pub success: bool,
    pub status: String,
    pub receipt_number: String,
    pub recipient_phone: String,
    pub amount_kes: i64,
    pub debited_usd: Decimal,
    pub new_balance_usd: Decimal,
    pub message: String,
}

pub fn ensure_mpesa_account(conn: &Connection, user_id: &str) -> String {
    let mut stmt = conn
        .prepare("SELECT id FROM accounts WHERE user_id = ? AND (name LIKE '%M-Pesa%' OR id = 'mpesa_wallet')")
        .unwrap();

    let account_id: Option<String> = stmt.query_row([user_id], |row| row.get(0)).ok();

    if let Some(id) = account_id {
        id
    } else {
        let new_id = format!("acc_mpesa_{}", &Uuid::new_v4().to_string()[..8]);
        conn.execute(
            "INSERT INTO accounts (id, user_id, name, type, account_class, currency, institution, initial_balance)
             VALUES (?, ?, 'M-Pesa Mobile Wallet', 'cash', 'asset', 'KES', 'Safaricom M-Pesa', 0)",
            params![new_id, user_id],
        )
        .unwrap();
        new_id
    }
}

pub fn initiate_stk_push(
    conn: &Connection,
    user_id: &str,
    phone_number: &str,
    amount_kes: i64,
    description: Option<&str>,
) -> Result<StkPushResult, String> {
    if amount_kes <= 0 {
        return Err("M-Pesa amount must be positive".to_string());
    }

    let mut clean_phone = phone_number.replace(|c: char| !c.is_numeric(), "");
    if clean_phone.starts_with('0') {
        clean_phone = format!("254{}", &clean_phone[1..]);
    } else if !clean_phone.starts_with("254") {
        clean_phone = format!("254{}", clean_phone);
    }

    let mpesa_account_id = ensure_mpesa_account(conn, user_id);
    let checkout_request_id = format!("ws_CO_{}_{}", Utc::now_timestamp_millis(), &Uuid::new_v4().to_string()[..8]);
    let merchant_request_id = format!("mr_{}", &Uuid::new_v4().to_string()[..12]);
    let rail_tx_id = format!("rail_{}", Uuid::new_v4());

    let desc = description.unwrap_or("Lumen Wallet Top-up");
    let converted_usd = (Decimal::from(amount_kes) * KES_TO_USD_RATE).round_dp(2);

    let meta = serde_json::json!({
        "description": desc,
        "merchantRequestId": merchant_request_id,
        "accountId": mpesa_account_id,
        "convertedUsd": converted_usd
    });

    conn.execute(
        "INSERT INTO payment_rails_transactions (id, user_id, rail, direction, status, amount, currency, reference, phone_or_account, metadata_json)
         VALUES (?, ?, 'mpesa', 'inflow', 'pending', ?, 'KES', ?, ?, ?)",
        params![rail_tx_id, user_id, amount_kes as f64, checkout_request_id, clean_phone, meta.to_string()],
    ).map_err(|e| e.to_string())?;

    Ok(StkPushResult {
        success: true,
        checkout_request_id,
        merchant_request_id,
        rail_tx_id,
        phone_number: clean_phone.clone(),
        amount_kes,
        status: "pending".to_string(),
        message: format!("STK push prompt dispatched to {}. Enter PIN on handset.", clean_phone),
    })
}

pub fn send_b2c_payout(
    conn: &mut Connection,
    user_id: &str,
    recipient_phone: &str,
    amount_kes: i64,
    description: Option<&str>,
) -> Result<PayoutResult, String> {
    if amount_kes <= 0 {
        return Err("Payout amount must be positive".to_string());
    }

    let mut clean_phone = recipient_phone.replace(|c: char| !c.is_numeric(), "");
    if clean_phone.starts_with('0') {
        clean_phone = format!("254{}", &clean_phone[1..]);
    } else if !clean_phone.starts_with("254") {
        clean_phone = format!("254{}", clean_phone);
    }

    let mpesa_account_id = ensure_mpesa_account(conn, user_id);
    let usd_amount = (Decimal::from(amount_kes) * KES_TO_USD_RATE).round_dp(2);

    let current_bal = get_account_balance(conn, &mpesa_account_id)?;
    if current_bal < usd_amount {
        return Err(format!("Insufficient funds in M-Pesa account. Available: ${}, Requested: ${}", current_bal, usd_amount));
    }

    let receipt_num = format!("B2C_{}", &Uuid::new_v4().to_string()[..10].to_uppercase());
    let rail_tx_id = format!("rail_{}", Uuid::new_v4());
    let desc = description.unwrap_or("Lumen Payout");

    let ledger_res = record_spending(
        conn,
        user_id,
        &mpesa_account_id,
        usd_amount,
        "M-Pesa Payout",
        &format!("M-Pesa Outward to {}", clean_phone),
        None,
        Some(receipt_num.clone()),
        "arrow-up-right",
        "orange",
    )?;

    let meta = serde_json::json!({
        "description": desc,
        "convertedUsd": usd_amount
    });

    conn.execute(
        "INSERT INTO payment_rails_transactions (id, user_id, rail, direction, status, amount, currency, reference, phone_or_account, metadata_json, journal_entry_id)
         VALUES (?, ?, 'mpesa', 'outflow', 'completed', ?, 'KES', ?, ?, ?, ?)",
        params![rail_tx_id, user_id, amount_kes as f64, receipt_num, clean_phone, meta.to_string(), ledger_res.journal_entry_id],
    ).map_err(|e| e.to_string())?;

    let new_bal = get_account_balance(conn, &mpesa_account_id)?;

    Ok(PayoutResult {
        success: true,
        status: "completed".to_string(),
        receipt_number: receipt_num,
        recipient_phone: clean_phone.clone(),
        amount_kes,
        debited_usd: usd_amount,
        new_balance_usd: new_bal,
        message: format!("KES {} sent to {} via M-Pesa B2C.", amount_kes, clean_phone),
    })
}

struct Utc;
impl Utc {
    fn now_timestamp_millis() -> u128 {
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_millis()
    }
}
